docs/performance-guide.md

```diff
@@ -25,150 +53,239 @@ guide intends to answer the following questions:
+
+The following formula is easily to be inferred based on the previous existing formula:
+
+```
+outboundConnections = outboundBandwidth * sendInterval / messageSize
+```
+
+For unit 100, we know the max outbound bandwidth is 400M from previous table,
+then for 20K message size, the max outbound connections should be 400M \* 5 / 20K =
+100,000, which matches the real value.
 
-As the connections increase, ASRS needs more CPU and memory to keep them, thus impacts the inbound or outbound capacity.
+* Mixed use cases
 
-The known bottleneck of performance is CPU, memory, and ASRS internal message queue. Different use cases may hit different bottleneck.
+The real use case typically mixes the 4 basic use cases together: **echo**, **broadcast**, **send to group**, or **send to connection**. The methodology used to evaluate the capacity is to divide the mixed use cases into 4 basic use cases, **calculate the maximum inbound and outbound message bandwidth** using the above formulas separetely, and sum them to get the total maximum inbound/outbound bandwidth. Then pick up the proper tier from the maximum inbound/outbound bandwidth tables.
 
-App server also impacts performance. Azure SignalR SDK default creates 5 server
-connections with ASRS. In the below performance test, server connections are
-increased to 15 (or more for broadcast and send message to big group) and
-StandardF4sV2 VM is chosen as app server. Different use cases have different
-requirement on app servers. Broadcast needs very small number of app servers.
-Echo or send to client needs many app servers.
+Meanwhile, for sending message to hundreds or thousand small groups, or thousands of clients sending message to each other, the routing cost will become dominant. This impact should be taken into account. More details is covered in the following "Case study" sections.
 
-This document summarizes 4 typical use cases for Websockets transport: echo, broadcast,
-send to group, and send to client, which will be introduced in the following
-sections.
+For the use case of sending message to clients, make sure the app server is **NOT** the bottleneck. "Case study" section gives the guideline about how many app servers you need and how many server connections should be configured.
 
-For all use cases, the performance guide wants to find the max inbound and
-outbound with the criteria that 99% message latency is less than 1s.
+<a name="case_study"></a>
+Case study
+----------
+
+The following sections go through 4 typical use cases for Websockets transport: **echo**, **broadcast**, **send to group**, and **send to connection**. For every scenario, it lists the current ASRS inbound and outbound capacity, meanwhile explains what is the main factors on performance.
+
+For all use cases, the performance guide wants to find the maximum inbound and outbound with the criteria that 99% message latency is less than 1 second.
```

Comment:
I think we should be emphasize this at the beginning of this article. (Maybe add a paragraph about our test methodology like network condition, how to measure latency, etc.)

Reply:
Ok

Reply:
Methodology introduction has added.